### PR TITLE
Implement activate/deactivate/delete for storage pools

### PR DESCRIPTION
**Pull request by KKoukiou**

### CODE REVIEW

**Comment by KKoukiou** [comment]
@Antique can you please take a quick look in libvirt API usage?

Also I am not sure regarding the correctness of one behaviour. Right now the delete operation will only succeed if the storage pool is empty. That means there is not volume in that pool. Is this ok or should I empty to content of the pool always when delete pool is selected.
There is another open PR to implement volume deletion btw.

**Comment by KKoukiou** [comment]
@Antique can you please take a quick look in libvirt API usage?

Also I am not sure regarding the correctness of one behaviour. Right now the delete operation will only succeed if the storage pool is empty. That means there is not volume in that pool. Is this ok or should I empty to content of the pool always when delete pool is selected.
There is another open PR to implement volume deletion btw.

**Comment by KKoukiou** [comment]
activate-pool

**Comment by KKoukiou** [comment]
deactivate-pool

File: pkg/machines/libvirt-dbus.js [hunk 1]
```diff
@@ -1123,4 +1127,9 @@ function updateNetworkIface({ domXml, networkMac, networkState, networkModelType
     return null;
 }
 
+export function storagePoolDelete(connectionName, objPath) {
+    return clientLibvirt[connectionName].call(objPath, 'org.libvirt.StoragePool', 'Delete', [Enum.VIR_STORAGE_POOL_DELETE_NORMAL], TIMEOUT)
+            .then(() => clientLibvirt[connectionName].call(objPath, 'org.libvirt.StoragePool', 'Undefine', [], TIMEOUT));
+}
```
Comment (Antique) [inline pkg/machines/libvirt-dbus.js]: It might be good idea to give user an option to only udefine storage pool or delete it completely including the storage.  In addition not all storage pools supports delete operation, currently it's only DIR, FS, NFS, LOGICAL, VSTORAGE and ZFS.

File: pkg/machines/components/storagePools/storagePool.jsx [hunk 2]
```diff
@@ -29,68 +29,129 @@ import {
...
+        ];
+
+        return (
+            <ListingRow rowId={idPrefix}
+                columns={cols}
+                tabRenderers={tabRenderers}
+                listingActions={<StoragePoolActions actionErrorSet={this.actionErrorSet.bind(this)} storagePool={storagePool} />} />
+        );
+    }
+}
```
Comment (Antique) [inline pkg/machines/components/storagePools/storagePool.jsx]: Just a note and suggestion, these changes looks like refactor of the old code.  For review and git history it's better to split refactoring and new code into separate patches to make it clear which part is only refactoring.

File: pkg/machines/components/storagePools/storagePool.jsx [hunk 3]
```diff
@@ -143,15 +150,26 @@ class StoragePoolActions extends React.Component {
...
         } else {
-            return deleteBtn;
+            deleteBtnWrapper = deleteBtn;
         }
+
+        return (
+            <React.Fragment>
+                { !storagePool.active &&
+                <Button id={`activate-${id}`} onClick={this.onActivate.bind(this)}>
+                    {_("Activate")}
```
Comment (Antique) [inline pkg/machines/components/storagePools/storagePool.jsx]: What was the reason to name this operation `Activate` instead of `Start`?

When we create storage pool in cockpit there is an option `Startup` with description `Start pool when host boots` so for consistency we should use `Start` here as well.

File: pkg/machines/components/storagePools/storagePool.jsx [hunk 4]
```diff
@@ -163,6 +170,10 @@ class StoragePoolActions extends React.Component {
 
         return (
             <React.Fragment>
+                { storagePool.active &&
+                <Button id={`deactivate-${id}`} onClick={this.onDeactivate.bind(this)}>
+                    {_("Deactivate")}
```
Comment (Antique) [inline pkg/machines/components/storagePools/storagePool.jsx]: Same here as for the `Activate` name, for consistency I would prefer `Stop` as opposite to `Start`.

File: pkg/machines/components/storagePools/storagePool.jsx (hunk 2 above)
Comment (KKoukiou) [reply pkg/machines/components/storagePools/storagePool.jsx]: And the diff is so bad because I had to change this component from stateless to statefull to support the errors.

File: pkg/machines/libvirt-dbus.js (hunk 1 above)
Comment (KKoukiou) [reply pkg/machines/libvirt-dbus.js]: Why only undefine the storage pool without deletion? One doesn't have to option to re-define existing storage anyway. One can only create one from scratch.

File: pkg/machines/components/storagePools/storagePool.jsx (hunk 3 above)
Comment (KKoukiou) [reply pkg/machines/components/storagePools/storagePool.jsx]: I made this decision according to the state wording. State is active/inactive not started/stopped. However I don't have strong opinion on it, I am ok to change this to match the other sentence you mentioned.

File: pkg/machines/components/storagePools/storagePool.jsx (hunk 4 above)
Comment (KKoukiou) [reply pkg/machines/components/storagePools/storagePool.jsx]: Same as above, I am ok to change it for consistency. But since we can change the other sentence as well, let's pick the better naming now that it's not really established.

@garrett what do you think?
Storage Pools start/stop vs activate/deactivate
I guess it's a subtle difference anyway.

File: pkg/machines/libvirt-dbus.js (hunk 1 above)
Comment (Antique) [reply pkg/machines/libvirt-dbus.js]: That is not true, cockpit is not calling `virStoragePoolBuild` libvirt API and when calling `virStoragePoolCreate` cockpit is not using any of the `build` flags so the storage pool is not created from scratch, it's just defining some existing path on the host.

So for example in case if DIR storage pool the delete operation removes the directory as well and build operation would create all necessary directories in order to use that storage pool.  The storage directory creation is not supported by cockpit, if you are creating new storage pool you need to provide existing directory therefore the `build` operation is not possible.

Counter-part to current storage creation in cockpit is only `undefine` from libvirt.  For that reason the `delete` option should be optional.

File: pkg/machines/components/storagePools/storagePool.jsx (hunk 2 above)
Comment (Antique) [reply pkg/machines/components/storagePools/storagePool.jsx]: Well :) changing component from stateless to statefull is the refactoring part and even though it's done in order to implement new functionality it should be in separate commit. It was just an suggestion based on good practices.

File: pkg/machines/libvirt-dbus.js (hunk 1 above)
Comment (KKoukiou) [reply pkg/machines/libvirt-dbus.js]: Good point that are not calling virStoragePoolBuild API. This has to be fixed, we need to support building storage pools as well. virt-manager allows that and we should here as well.

On the other hand how would you imagine the UI to allow undefining pools without deleting them? virt-manager doesn't offer this option from what I see.
And what about the storage volumes they contain. Should we force the user to remove first all volumes and then attempt pools deletion of should be force delete?

**Comment by KKoukiou** [comment]
A bit of backround
-----------------------

Initially I wanted to implement deletion of storage pools (instead of just undefining them), which would mean deletion of the content of a pool and then undefining it from libvirt.

However, deletion is not really straightforward with the current tools due to the following reasons.

- Libvirt requires deactivating a pool before using the Delete and/or Undefine APIs.
- Delete API works only for nfs and local fs  pools types and for these it performs just a plain rmdir. Which will fail if the pool is not empty. So we would have to manually remove the pool contents by removing the volumes inside the pool and then attempting deletion.
- But ListVolumes API works only for activated pools so even if we wanted to fetch the volumes of the pools we are about to delete this wouldn't be possible from libvirt.

To sum up, removing the content of pools before attempting deletion is an procedure that can't be easily automated, virt-manager doesn't perfom deletion either (it just undefines them).
For the time being, I think it's better to not support deletion of the content of the pools. However the volume deletion is going to be implemented separately.

File: pkg/machines/components/storagePools/storagePool.jsx [hunk 5]
```diff
@@ -29,68 +29,129 @@ import {
...
+        const id = storagePoolId(storagePool.name, storagePool.connectionName);
+        const undefineBtn = (
+            <Button id={`undefine-${id}`} bsStyle='danger' disabled={storagePool.active} onClick={this.onUndefine.bind(this)}>
+                {_("Undefine")}
+            </Button>
+        );
+
+        if (storagePool.active) {
+            return (
+                <OverlayTrigger overlay={ <Tooltip id='tip-pool-undefine'>{_("Try deactivating the storage pool before attempting to undefine it")}</Tooltip> } placement='top'>
```
Comment (martinpitt) [inline pkg/machines/components/storagePools/storagePool.jsx]: There is no button to deactivate the pool, right? 

**Update**: The following commit adds it.

**Comment by KKoukiou** [comment]
activate-storage-pool
deactivate-storage-pool

File: pkg/machines/components/storagePools/storagePool.jsx [hunk 6]
```diff
@@ -128,6 +128,13 @@ StoragePool.propTypes = {
 };
 
 class StoragePoolActions extends React.Component {
+    onActivate() {
+        storagePoolStart(this.props.storagePool.connectionName, this.props.storagePool.id)
```
Comment (martinpitt) [inline pkg/machines/components/storagePools/storagePool.jsx]: Let's please use a consistent naming. If the libvirt terminology is Activate/Deactivate, then use that everywhere, othwise Start/Stop.

File: pkg/machines/components/storagePools/storagePool.jsx (hunk 3 above)
Comment (martinpitt) [reply pkg/machines/components/storagePools/storagePool.jsx]: Please no `.bind()` call in `render()`, do this in the constructor instead to only do it once instead of once per rendering.

File: pkg/machines/components/storagePools/storagePool.jsx (hunk 6 above)
Comment (KKoukiou) [reply pkg/machines/components/storagePools/storagePool.jsx]: virsh commands are ```virsh pool-start```` and ```virsh pool-destroy````.
The API exposes the virStoragePoolCreate and virStoragePoolDestroy function for the above.
And then the libvirt documentation uses active/inactive storage pool to refer to their state.
So there is some mess with naming from libvirt side already :(

But yes, let's keep the function names at least consistent, I think I prefer activate/deactivate but I am open to suggestions.

**Comment by KKoukiou** [comment]
After rethinking, I actually implemented the delete operation, for active storage pools. The user will be promted to delete the pool's volumes, before deleting an active pool.
The default is however to not delete them, since many pool types do not support deleting their volumes. (ex iscsi)
When trying to delete an active storage pool, in the background we have to deactivate it first.

In the future the dialog can be enhanced to guess if the pool type supports deletion, but for the time being I think this is good in matters on functionality.

delete-active-pool


delete-inactive-pool

**Comment by KKoukiou** [comment]
Just clean rebase.

File: pkg/machines/components/storagePools/storagePoolDelete.jsx [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+        }
+    }
+
+    render() {
+        const { storagePool } = this.props;
+        const id = storagePoolId(storagePool.name, storagePool.connectionName);
+
+        let defaultBody = (
+            <div className='ct-form-layout'>
+                { storagePool.active && storagePool.volumes && <React.Fragment>
```
Comment (martinpitt) [inline pkg/machines/components/storagePools/storagePoolDelete.jsx]: Should also test for `volumes.length > 0` here? (empty lists are truthy, irritatingly)

File: test/verify/check-machines-dbus [hunk 8]
```diff
@@ -235,6 +235,40 @@ class TestMachinesDBus(machineslib.TestMachines):
...
+        # Delete and active Pool and also it's volumes
+        b.wait_present("#delete-pool-myPoolTwo-{0}".format(connectionName))
+        b.click("#delete-pool-myPoolTwo-{0}".format(connectionName))
+        b.wait_present("div.modal-dialog")
+        b.wait_present("input#storage-pool-delete-volumes")
+        b.set_checked("input#storage-pool-delete-volumes", True)
+        b.click('div.modal-footer button:contains("Delete")')
+        b.wait_not_present("div.modal-dialog")
+        b.wait_not_present("#pool-myPoolTwo-{0}-storage-volumes-list".format(connectionName))
+        self.assertIn("Empty", m.execute('[ "$(ls -A /mnt/vm_two)" ] || echo Empty').strip())
```
[the remark this answers is not in the document]
Comment (KKoukiou) [reply test/verify/check-machines-dbus]: Actually we don't delete the directory, just it's contents. I am not sure if this is the behaviour we want to keep for the future, but deleting the storage pool itself is not always straightforward as it's not always a directory.

File: pkg/machines/libvirt-dbus.js [hunk 9]
```diff
@@ -1144,4 +1148,13 @@ export function storagePoolDeactivate(connectionName, objPath) {
     return call(connectionName, objPath, 'org.libvirt.StoragePool', 'Destroy', [], TIMEOUT);
 }
 
+export function storagePoolUndefine(connectionName, objPath) {
+    return call(connectionName, objPath, 'org.libvirt.StoragePool', 'Undefine', [], TIMEOUT);
+}
+
+export function storageVolumeDelete(connectionName, path) {
+    return call(connectionName, '/org/libvirt/QEMU', 'org.libvirt.Connect', 'StorageVolLookupByPath', [path], TIMEOUT)
+            .then(objPath => call(connectionName, objPath[0], 'org.libvirt.StorageVol', 'Delete', [0], TIMEOUT));
```
[the remark this answers is not in the document]
Comment (KKoukiou) [reply pkg/machines/libvirt-dbus.js]: It can fail but I have the fail handler where I call this function and I display the error in the UI.
Just for verification I replaced the method name from StorageVolLookupByPath to StorageVolLookupByPath-test and tried to delete a pool and the error I see is the following.
vollookupbypath-fail

File: pkg/machines/components/storagePools/storagePoolDelete.jsx (hunk 7 above)
Comment (KKoukiou) [reply pkg/machines/components/storagePools/storagePoolDelete.jsx]: Technically this will not cause problems because I don't add the volumes object at all if a pool doesn't have volumes. So if it exists it's always populated.
But I 'll replace it, it's not clear code indeed.

File: test/verify/check-machines-dbus [hunk 10]
```diff
@@ -235,6 +235,40 @@ class TestMachinesDBus(machineslib.TestMachines):
...
+
+        # Delete an inactive Pool. It's volumes won't be deleted
+        b.wait_present("#delete-pool-myPoolTwo-{0}".format(connectionName))
+        b.click("#delete-pool-myPoolTwo-{0}".format(connectionName))
+        b.wait_present("div.modal-dialog")
+        b.wait_in_text("div.modal-dialog div.modal-body div.ct-form-layout",  "Its content will not be deleted.")
+        b.click('div.modal-footer button:contains("Delete")')
+        b.wait_not_present("div.modal-dialog")
+        b.wait_not_present("#pool-myPoolTwo-{0}-storage-volumes-list".format(connectionName))
+        self.assertNotIn("Empty", m.execute('[ "$(ls -A /mnt/vm_two)" ] || echo Empty').strip())
```
[the remark this answers is not in the document]
Comment (KKoukiou) [reply test/verify/check-machines-dbus]: Ah, I didn't check the return code of ls. I put it inside $() so I checked the output of it which is empty for empty dirs otherwise some string. And inside a conditional check of [] this return 0 or 1 like this.
``` bash
$ [ 'test' ] || echo Empty
$ [ ] || echo Empty
```
But I agree your command is cleaner ;) 
Empty

File: test/verify/check-machines-dbus (hunk 8 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply test/verify/check-machines-dbus]: OK, then once again the `|| echo Empty` would never run.

[the remark this answers is not in the document]
Comment (martinpitt) [reply test/verify/check-machines-dbus]: Ah, nevermind, it's wrapped in yet another `[ ]`. Really not that easy to read :-)

**Comment by martinpitt** [comment]
Semaphore fails:

    test/verify/check-machines-dbus:247:77: E241 multiple spaces after ','

Otherwise looks good now, thanks!

**Comment by KKoukiou** [comment]
Activate pool
activate-storage-pool
Deactivate pool
deactivate-storage-pool
Delete inactive pool
delete-inactive-pool
Delete active pool
delete-active-pool
